### PR TITLE
Rewrite timing logic for updating delay set

### PR DESCRIPTION
**Pull request by holm**
This rewrites the logic around updating delay set to be a bit safer and simpler (imo).

In the previous code there was a race-condition due to the timer being cleared before the call to process the delay set, and in the time until the processing was done, another processing could have been started.

The new code does away with the guardian timer, and instead combines the guardian timer with the updates coming from new jobs being posted.

Refs https://github.com/OptimalBits/bull/issues/1222

### CODE REVIEW

**Comment by manast** [comment]
thank you for your attempt, I tried myself to reduce the complexity of the code before, however it is quite difficult to fix all the edge cases, for instance with this solution most repeatable job tests are failing. As a side note you may be interested in the new implementation for bull 4, however this requires redis 5.0 and above: https://github.com/taskforcesh/bull/pull/1/files#diff-e43f718fe879603152e9c16fba64d4cfR49

**Comment by holm** [comment]
This seems to just be due to the way the tests are written. I can change the code a little bit to make the tests pass as they are.

**Comment by holm** [comment]
Should be all green now

File: lib/queue.js [hunk 1]
```diff
@@ -851,49 +814,38 @@ Queue.prototype.run = function(concurrency) {
...
+Queue.prototype.updateDelayTimer = function() {
+  return scripts
+    .updateDelaySet(this, Date.now())
+    .then(nextTimestamp => {
+      this.delayedTimestamp = nextTimestamp
+        ? nextTimestamp / 4096
+        : Number.MAX_VALUE;
+
+      // Clear any existing update delay timer
+      if (this.delayTimer) {
```
Comment (manast) [inline lib/queue.js]: this if is not needed since clearTimeout(0) is a noop

**Comment by manast** [comment]
good job. I will merge this but I will wait a couple of days before doing a release so that you can run this in production and check if it works better than the previous version in your scenario.

**Comment by holm** [comment]
Thanks! I will update our production environment tomorrow and report back what I see there.

**Comment by gcox** [comment]
@holm @manast This has been working well for me for about 12 hours now.

**Comment by holm** [comment]
@gcox That is good to hear. Did you have problems before, and what were they exactly?

**Comment by gcox** [comment]
@holm Yes, I suffered from #1226 as well as #1222

#### Setup
A rate limited queue (1 job every 3 seconds), with about 1000 jobs.

#### Issues

**Excessive bandwidth** - Bandwidth for my redis instance was averaging 5Mbps in & out. Interestingly, in 3.6, it was averaging 5Mbps in & 1.5Mbps out. In any case, with your fix, it's now averaging 0.25Mbps in & <1Mbps out.

**High CPU** - Once a worker was started, CPU usage increased exponentially. After a couple of minutes it was at 100%. Now it's back to a consistent low usage.

I believe both of those issues were made worse due to using the rate limiter.

---

So, thanks for your fix 👏

**Comment by holm** [comment]
Thats great, thanks for the update. 

I must admit I find queue logic really hard to get right, so I would definitely keep an eye on the jobs and make sure all the delayed and rate limited jobs gets properly processed, just to be on the safe side :)

File: lib/queue.js [hunk 2]
```diff
@@ -851,49 +814,38 @@ Queue.prototype.run = function(concurrency) {
...
+      ]);
 
-    const update = delay => {
-      if (delay) {
-        delay = delay < MAX_TIMEOUT_MS ? delay : MAX_TIMEOUT_MS;
-        this.delayTimer = setTimeout(delayUpdate, delay);
+      // Schedule next processing of the delayed jobs
+      if (delay <= 0) {
+        // Next set of jobs are due right now, process them also
+        this.updateDelayTimer();
```
Comment (gabegorelick) [inline lib/queue.js]: I'm getting warnings about unhandled promises stemming from this. It seems like this should return here?

Comment (gabegorelick) [reply lib/queue.js]: Nevermind, I see `updateDelayTimer` handles errors itself. I must be running into something else.
